### ISSUE TITLE
Suggestion: uri scheme

### ISSUE THREAD

**Issue by evert** [google/open-location-code#248]
An identification system can benefit from a uri scheme. Other examples:

    mailto:example@example.org
    tel:+1555445323
    geo:43.653349,-79.398685
    urn:ISBN:0-395-36341-1

Has this been considered for this project? I think it would be a great addition to make it work in various other systems that are already enabled to work with uri's.

**Comment by zongweil** [comment]
Can you expand on the potential benefits that you see? One of the advantages of plus codes is that you can identify them from a random string without needing to be told that it's a plus code via a URI. Not sure if that's what you were referring to or if there were other benefits you had in mind.

@drinckes to comment if any thought has been put into this

**Comment by evert** [comment]
One advantage is that, as a uri, it can appear in existing formats that already take uri's, such as `<link>` elements in HTML, RFC / JSON-LD, or basically any web technology that machines can parse.

Or is this project strictly intended for reading by humans / unstructured text? Would you recommended to always convert to the `geo:` uri schema for machine-to-machine communication?

Worth noting that when written by humans, `mailto:` and `tel:` is also typically not used. The uri scheme is only used in contexts where they are really needed.

**Comment by WilliamDenniss** [comment]
I think it makes sense to use the existing `geo:` (RFC5870) scheme for this use-case, as that scheme can already indicate a location accurately. In HTML, the link could be geo:, while the text could be a plus code.

I wouldn't say plus codes are *strictly* for reading by humans and use in unstrucuted text, but that is certainly where a lot of the value is added over decimal degrees.

**Comment by drinckes** [comment]
The problem with using the existing `geo` URI is that it is already defined to be `lat,lng[,altitude]`. Changing the definition wouldn't be trivial - and then you have to change all the things that accept `geo` URIs.

Like you point out @evert , plus codes big advantage is when location is being communicated to, from or between people. URIs by their nature are generated by a computer, and interpreted by a computer. `geo` URIs don't support arbitrary search strings, or addresses, so 

For example, if I get a `geo` URI for a location, click on it, and my maps app opens at that location - I'm going to want that app to work out the plus code and show it to me, just like it does the street address.

**Comment by WilliamDenniss** [comment]
To clarify, when I said "it makes sense to use the existing geo: scheme for this use-case", I meant that I believe the `geo:` scheme, as it is defined today, already solves this use-case (i.e. without needing plus codes).

Agree that an ideal UX would be to open the maps app, and for that app to show the plus code for easy sharing and communication (just like for street addresses, as you say).

This discussion (and others I've had) makes me wonder if the OLC definition should include some explicit text on non-goals, e.g. it's a non-goal to replace the decimal degrees representation in all situations.

**Comment by Erudition** [comment]
> URIs by their nature are generated by a computer, and interpreted by a computer. 

But many URI types, while meant for machines to interpret in the end, are specifically designed to be nice to transfer human-human if possible. Otherwise we could all just use IP addresses to navigate to websites - if we'd only ever get a URL from a computer, which we could then just click on! Instead, we have a global address (like google.com) that you can easily tell to someone over the phone, or if you didn't know how to copy/paste, or any number of other scenarios.

In fact I think that's a good example. Plus codes are trying to be, to Lat/Lon coordinates, what the human-readable domain-name-based URLS like google.com/maps are to IP-address-based URLs. I'd say dealing with raw IP addresses, is nearly as not-human-friendly as dealing with coordinates instead of addresses. 

> it's a non-goal to replace the decimal degrees representation in all situations

Maybe not all, but the proposed is a pretty big situation. If I want to send you a link _right now_, I just put it in markdown as `[google](google.com)`  and you'll be able to click on it - [google](google.com) - and have it **instantly interpreted and opened in your browser**.  If I fail to do so, you'd just get the text google.com, which you'd then have to copy and paste (or temporarily memorize, then type manually) in your browser to see the referred resource. That's annoying. We're all citizens of the web and know how lazy people can be - hyperlinking that text, or not, can be the difference in whether someone actually follows the link or not. And if they're a technological novice, there are many more good reasons than laziness.

Think similarly about plus codes. If I simply say 87G7MXQ4+M5, a reader would have to
- Know that this is in fact a code, not just a random string (context may be enough)
- Identify this as a plus code, or at least a way of identifying a location (a very small portion of the population can do this currently)
- Know what services or apps are relevant for interpreting this datum (e.g. Google Maps) and how to use one of them (even smaller)
- Know how to transfer the location to one of those services (e.g. copy/paste, or richer interfaces)

before being shown the location. Note that this list is pretty similar for raw coordinates, except the likelihood of each step is far greater. But like URLS, if I simply (and easily!) throw a `geo:` in front of those coordinates, you don't need ANY of those steps to be successful. Even someone who has no idea what they're clicking will soon realize it's a location - and at the same time, have all the relevant cues to know exactly what that location is! 

So if we can just specify a location like this: [geo:43.653349,-79.398685](geo:43.653349,-79.398685)
**Then why not make it just as easy (for humans!) in the encoding that's meant for humans?** 

So a few extra characters of typing, here on github for example, would turn 87G7MXQ4+M5 into [plus:87G7MXQ4+M5](https://plus.codes/87G7MXQ4+M5). Just gotta throw a `plus:` in front of it! (Or a `geo:`, if we feel like extending that spec). A tiny effort for a crazy high boost in usability.

Here's an example of how that may one day work: 

[87G7MXQ4+M5](https://plus.codes/87G7MXQ4+M5)

Clicking on a `plus:` link, the worst that can happen is that you have no local app that knows how to handle location URIs, and it fails. But many people do - in fact, you may even have multiple apps that can handle that link, and an OS like Android may ask you to  choose which one you'd prefer to use! Nice.

I encoded that link in this way:

> In HTML, the link could be geo:, while the text could be a plus code.

...but I shouldn't have to - that process forfeits the benefits described above. I'd have to take steps to resolve the Plus code into coordinates, which is again asking a lot more from people, and increasing the barrier to entry for making clickable plus codes.

And lets not assume all of our software will just get smart and make all plus codes clickable automatically - it's been this long and we still only have that for URLs in some places but not others (let alone clickable coordinates!). We'd be requiring every app ever to start parsing for an additional thing instead of just parsing ordinary URIs with protocols. (Sure, the `+` in the plus code may make it easier for _humans_ to identify, but for machines that doesn't make much of a difference in the workload required.)

**Comment by bocops** [comment]
I think we need to be precise about the various use cases we're talking about here.

One use case is that I want to let someone know about a location in a mostly **non-digital** context. Perhaps I'm contacting you over phone and want to tell you where we can meet in half an hour: "I'm sitting in the park at CFGH+23, from there it's only a five minute walk to a great cafe!". Perhaps I'm writing a letter to someone, and know that this person can be reached by addressing the letter to "QXV4+C2, CITY_NAME".

In either case, the fact that this string is something address-like is context information. Adding a URI prefix is not necessary, and might not even be useful: Just like we often omit parts of web addresses if they are considered to be shared knowledge or redundant (it's really "HTTPS colon slash slash WWW dot google dot com", but we typically say "google com" or even just "google"), having a "plus:" prefix that doesn't even hint at the meaning of the string that follows could be considered redundant information.

A different use case is to let someone know about a location while **both parties have access to some digital device with certain capabilities**. In that case, instead of using a mapping app on my device to get a plus code for my current location, sending it to you in a URI format that still needs to be defined, and finally having your device decipher that format and showing it to you - it's probably easier to just share a location via Google Maps or, more generally speaking, use any other existing encoding because it's not necessary for the data to be in plus code format for this to work.

That seems to leave use cases where the data needs to be both human-readable and automatically turned into a link by the software that displays the human-human conversation. This would definitely be useful in our current context of talking about the underlying technology itself. I wonder if it would really be useful in any other context, though (like a random messenger conversation), because we'd quickly be back to the previous use case.

Are there any other cases I'm not seeing?

**Comment by Erudition** [comment]
> One use case is that I want to let someone know about a location in a mostly **non-digital** context.

No, because this issue is about URIs, and you wouldn't use a URI for non-digital communication. You'd never say to someone in person "my email address is mailto:me@domain.com", and that's not what it's for.

> having a "plus:" prefix that doesn't even hint at the meaning of the string that follows

- This absolutely hints at the meaning of the string that follows, for those that are familiar with plus codes. This is one of the reasons for "+" being in plus codes.
- Being self-explanatory (like `tel:`) is nice, but not required (as in the `urn:` example). 
- For those unfamiliar with the concept of plus codes, this argument might hold water, but in that case it's really just an argument to use `geo` instead. That's fine too, just harder.
- If both parties are unfamiliar with plus codes, neither would be using it's protocol URIs.
- another form of "context" is what happens when you click it. These links aren't meant to be read, they're meant to be clicked. Clicking a plus code and observing the result should be a pretty effective way of learning that it means a location. 
- If they are reading the URI in non-link form for some reason, the prefix can be safely discarded without additional confusion. Just like people who accidentally put the `mailto:` in the visible part of a link rather than just the underlying URI.


Regarding your second use case,

> it's probably easier to just share a location via Google Maps

Of course you can, but this misses the point of defining a standard. Google Maps is a proprietary application and we would like to not have to rely on the recipient having it to preview a geographic location. Now as far as it being "easier", it should be just as easy for either scenario when we're done.

 instead of using a mapping app on my device to get a plus code for my current location, sending it to you in a URI format, and finally having your device decipher that format and showing it to you

> that still needs to be defined

well, right now, sure, but that's what we're hoping to change. Using `mailto` is perfectly reasonable practice as it is widely implemented, and it's not easy to to determine (and rude to assume) the software set up of your target audience - e.g. If I instead sent you a link to the compose screen of Gmail (because everybody uses gmail, right?) with my address pre-filled.

>  using a mapping app on my device to get a plus code for my current location, sending it to you in a URI format, and finally having your device decipher that format and showing it to you

Yup, that's what happens with `geo:` currently, if you spelled it out. Oh and also Google Maps links fit that process too (though they're much bulkier).

> [easier to] use any other existing encoding because it's not necessary for the data to be in plus code format for this to work.

So we shouldn't add support a feature because it's not yet supported?

> I wonder if it would really be useful in any other context, though (like a random messenger conversation) [...] we'd quickly be back to the previous use case.

Why? I often chat with people about, e.g. 22 Vasar Street. Plus codes are meant to be addresses, especially for places that don't have one otherwise. So it's safe to say that we should allow for a future where people use them just as casually - perhaps even knowing them by heart.

**Comment by bocops** [comment]
@Erudition I was talking about use cases of plus codes, not use cases of a URI scheme for plus codes.

The fact is that plus codes are deliberately designed for contexts that don't involve computers at most (if not all) steps. Whenever a plus code is written down on an envelope for another human to read, or spelled out over the phone for another human to transcribe, adding a URI prefix is unnecessary.

Even in those cases where a computer is involved, it is mostly to support one of the above "human-human interactions". I can read a plus code off Google Maps - or any other app that implements plus code generation - and I can type a plus code I received into the same app to see a map with a pin in the correct location.

A URI scheme would only be sensible
- if the code is itself transferred via digital text (say, in an e-mail or a forum post like this)
- AND if it is necessary for the other party to click on it (for example, most codes used here on GitHub are just examples that other participants don't need to resolve)
- AND if that can't be achieved any other way (like just prefixing with `https://plus.codes/` instead, or having a browser plugin that recognizes plus codes in their current format)
- AND if it is important to use plus codes instead of any other geo format.

This combination of restrictions is why I don't see any really important case for a URI prefix.

**Comment by Erudition** [comment]
Okay. Well I was talking about the use cases for a URI scheme for plus codes, because that is what this issue is about. 

> The fact is that plus codes are deliberately designed for contexts that don't involve computers at most (if not all) steps.

Well, they're deliberately designed to be _compatible_ with situations that don't involve computers, yes. Theoretically, you are able to whip out a specially printed map to choose plus codes for locations and to resolve codes from others. No tech required, let alone internet. That was indeed a design goal.

... In practice, though, most people will just whip out their device. Not just for some small part of the chain, but for most or all of it (choosing, encoding -> decoding, resolving). In the overwhelming majority of cases. Sure, maybe the inner part of the chain, of conferring the code to another human, might be non-digital, like a voice conversation instead of a text message - though even then, afterwards I'd send my buddy a text with the code I told him just so he doesn't need to memorize it (and so he can copy and paste into maps) just like with normal addresses. The non-computer use cases are certainly possible to think of, but they're hardly majority. And the future will only have more technology.

> A URI scheme would only be sensible
> This combination of restrictions is why I don't see any really important case for a URI prefix.
 
It's one thing to think of this feature request as not "_really important_", but saying it's not _sensible_ is another thing entirely. Instead of a list of bullets of things that would make the URI scheme "make sense", your list reads more like reasons a URI scheme would be absolutely mandatory - so of course it doesn't look convincing. For example, you have in there that

> it is necessary for the other party to click on it

Of course it will never be _"necessary"_ for the other party to click on it - as with right now, they could _always_ just copy the code and paste it into the maps app manually. But this issue isn't about proving the feature is _necessary_, it's about proving that it would be _beneficial_.

> if that can't be achieved any other way (like just prefixing with https://plus.codes/ instead, or having a browser plugin that recognizes plus codes in their current format)

Again, of course that could be achieved some other way! Why should this be a restriction? I'm glad you brought up those two workaround examples, because those (1. having the sender, rather than the recipient, decide the displaying medium, and likely relying on some centralized service like plus.codes to do so and 2. requiring an entire browser extension (!) just to display a link the same way a standard `geo:` URI would with no additional software) are good examples of the kind of nonsense we're trying to avoid.

> AND if it is important to use plus codes instead of any other geo format.

Wait, so are you actively discouraging the use of plus codes in general? I thought we could all at least agree here that our goal is to _promote_ the spread and adoption of plus codes (regardless of transfer medium) i_n general_, not to discourage it (even for only a specific medium).

**Comment by drinckes** [comment]
Creating a new URI is a lengthy process, and creating it doesn't mean that it will do anything (e.g. to date, neither Chrome nor Firefox desktop browsers handle the geo: URL).

I can't think of a use case for a plus code URI that is not already covered with the geo: URI. If you have an example use case, feel free to explain it, but unless it's overwhelming (or someone feels strongly enough about it to be willing to go through the URI process) I can't see this progressing.

**Comment by bocops** [comment]
> Wait, so are you actively discouraging the use of plus codes in general?

No. What I _want_ to discourage, though, is any use of plus codes just for the sake of it. The plus code format is basically just a wrapper around another format that is readily understood by our technology, so it doesn't make sense to use plus codes there unless there's some _other_ problem that gets solved by doing so. This other problem could be "easy human communication", but probably not "machine readability".

I also feel that I'm not the only one stating exactly that. Other than that, I think we've exchanged all our arguments and there's no need to continue rehashing them.

**Comment by Erudition** [comment]
> What I want to discourage, though, is any use of plus codes just for the sake of it.

I'm not sure what that means, other than "in general". If you want to encode location, you're gonna pick a format. So it just sounds like we disagree on whether plus codes should be used outside of cases where they're the only way:

> it doesn't make sense to use plus codes there unless there's some other problem that gets solved by doing so

Why doesn't it? Again, maybe it's because I see plus codes as more or less a replacement for geo coordinates; an answer to its deficiencies, such as being hard to parse by humans. In that case, anywhere a big ol' geo coordinate is used, would be _better served_ with a plus code. In each way it's either better ("easy human communication") or pretty much equivalent ("machine readability"). 

I'm sure some people have the idea that plus codes should be only used when necessary, surgically replacing only those cases when there's a clear advantage - as seems to be your take - but based on the majority of evidence (e.g. Google Maps implementing the plus code display on _all_ POI, _even ones that already have a perfectly good classic address_) I assumed that the project was more ambitious than that.

I guess it all comes down to what @WilliamDenniss said earlier:

> This discussion (and others I've had) makes me wonder if the OLC definition should include some explicit text on non-goals, e.g. it's a non-goal to replace the decimal degrees representation in all situations.

Perhaps not quite in _all_ situations, but in the context of _succeeding where classic geo coordinates failed_ (from the manifesto), I'd certainly hope that is a goal, or at least a measurement of success. Kind of like how tablets existed but never really 'took off' until the iPad (a better-marketed, but otherwise incremental improvement).

**Comment by fulldecent** [comment]
Recommending to close issue.

There may be interest for somebody to open an IETF RFC and propose a new top-level namespace for plus codes.

But that surely is outside the our scope of 

> a library to generate short codes, called "plus codes", that can be used as digital addresses where street addresses don't exist.

I am citing this project scope from the GitHub repository description at https://github.com/google/open-location-code under ABOUT on the side bar.

**Comment by rokejulianlockhart** [comment]
This is really disappointing. Since the developers of the official tools for a standard ultimately decide how a standard should be used due to their consequent influence, this should be reconsidered.

Additionally, to suggest that the reporter submit an RFC to the IETF to request designation of a top-level namespace themself is somewhat disingenuous, since surely such a process is monumentally more difficult for a regular person to pursue than Google's engineers.

**Comment by bocops** [comment]
@rokejulianlockhart That was not a decision, but just another comment by a regular participant in this conversation.

@bilst and @drinckes are the only "official" people (as in, "have worked at Google at the time") in here, but I've not seen any activity from either of them in probably two years.

**Comment by RokeJulianLockhart** [comment]
We don't actually have to take this issue to the IETF. For instance, https://security.stackexchange.com/q/272885/217497 demonstrates that a group can simply standardize around a collectively decided URN namespace. It's quite common. For instance, how about we use `URN:OLC`? https://www.google.com/search?q=%22urn%253Aolc%22 doesn't indicate any prior usage.
